ext/opentelemetry-ext-wsgi/src/opentelemetry/ext/wsgi/__init__.py

```diff
@@ -0,0 +1,105 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+The opentelemetry-ext-wsgi package provides a WSGI middleware that can be used
+on any WSGI framework (such as Django / Flask) to track requests timing through
+OpenTelemetry.
+"""
+
+import functools
+import wsgiref.util as wsgiref_util
+
+from opentelemetry import trace
+from opentelemetry.ext.wsgi.version import __version__  # noqa
+
+
+class OpenTelemetryMiddleware:
+    """The WSGI application middleware.
+
+    This class is used to create and annotate spans for requests to a WSGI
+    application.
+
+    Args:
+        wsgi: The WSGI application callable.
+    """
+
+    def __init__(
+            self,
+            wsgi,
+            span_context_propagator=None,
+            distributed_context_propagator=None,
+    ):
+        self.wsgi = wsgi
+
+        # TODO: implement context propagation
+        self.span_context_propagator = span_context_propagator
+        self.distributed_context_propagator = distributed_context_propagator
+
+    @staticmethod
+    def _add_request_attributes(span, environ):
+        span.set_attribute("component", "http")
+        span.set_attribute("http.method", environ["REQUEST_METHOD"])
+
+        host = environ.get("HTTP_HOST") or environ["SERVER_NAME"]
+        span.set_attribute("http.host", host)
+
+        url = (
+            environ.get("REQUEST_URI") or
+            environ.get("RAW_URI") or
+            wsgiref_util.request_uri(environ, include_query=False)
+        )
+        span.set_attribute("http.url", url)
+
+    @staticmethod
+    def _add_response_attributes(span, status):
+        status_code, status_text = status.split(" ", 1)
+        span.set_attribute("http.status_text", status_text)
+
+        try:
+            status_code = int(status_code)
+        except ValueError:
+            pass
+        else:
+            span.set_attribute("http.status_code", status_code)
+
+    @classmethod
+    def _create_start_response(cls, span, start_response):
+        @functools.wraps(start_response)
+        def _start_response(status, response_headers, *args):
+            cls._add_response_attributes(span, status)
+            return start_response(status, response_headers, *args)
+
+        return _start_response
+
+    def __call__(self, environ, start_response):
+        """The WSGI application
+
+        Args:
+            environ: A WSGI environment.
+            start_response: The WSGI start_response callable.
+        """
+
+        tracer = trace.tracer()
+        method = environ["REQUEST_METHOD"]
+        path = environ["PATH_INFO"]
+
+        span_name = "[{}]{}".format(method, path)
+
+        with tracer.start_span(span_name) as span:
+            self._add_request_attributes(span, environ)
+            start_response = self._create_start_response(span, start_response)
+
+            for yielded in self.wsgi(environ, start_response):
+                yield yielded
```

Comment:
Interesting, first time I've seem `yield yielded` :)

Reply:
I still vote to at least use `yield from`.

Reply:
While `yield from` is equivalent in this circumstance since wsgi applications are not coroutines, unfortunately using `yield from` doesn't solve the issue with WSGI applications becoming generators / deferring the call to the wsgi callable.

Reply:
I've been putting some thought into this and I think `yield from` would actually be incorrect in this case. While the call to `self.wsgi` return an iterable, that iterable is not guaranteed to be a generator so delegating `send` and `close` calls to the result of `self.wsgi` would be incorrect.

Note that `yield from` defined in [PEP 380](https://www.python.org/dev/peps/pep-0380/) defines `yield from` as being used by a generator to delegate part of its operations to another generator.

Since `self.wsgi` does not always return a generator, using `yield from` here would be incorrect.

Reply:
`yield from` supports, but does not require the iterable (it is defined as `yield from <iterable>´, not `yield from <generator>`, even though that seems to have been the main motivation) to implement close or throw or send. But the WSGI spec explicitly calls out `close` and generators, see https://www.python.org/dev/peps/pep-3333/#specification-details. So delegating `close` (if it exists) would actually be beneficial.

Quoting PEP 3333:
> If the iterable returned by the application has a close() method, the server or gateway must call that method upon completion of the current request, whether the request was completed normally, or terminated early due to an application error during iteration or an early disconnect of the browser. (The close() method requirement is to support resource release by the application. This protocol is intended to complement PEP 342's generator support, and other common iterables with close() methods.)

What is more, `yield from` seems to be a tiny bit faster than a manual loop:

```python
In [1]: def generate(xs): yield from xs

In [4]: def generate2(xs):
   ...:     for x in xs:
   ...:         yield x
   ...:

In [6]: xs = [b'<html>'] * 1024

In [18]: %timeit for _ in generate(xs): pass
23.5 µs ± 374 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)

In [19]: %timeit for _ in generate2(xs): pass
30.2 µs ± 729 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)


In [21]: xs = xs[:1]


In [22]: %timeit for _ in generate(xs): pass
196 ns ± 1.19 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)

In [23]: %timeit for _ in generate2(xs): pass
202 ns ± 4.79 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)
```